services/ftp/ftp_service.py

```diff
@@ -0,0 +1,173 @@
+
+    def on_disconnect(self):
+        self.__send_alert(CLIENT_DISCONNECTED_DESCRIPTION)
+
+    def on_login(self, username):
+        self.__send_alert(USER_LOGIN_DESCRIPTION)
+
+    def on_login_failed(self, username, password):
+        self.__send_alert(USER_FAILED_LOGIN_DESCRIPTION, {
+            USERNAME: username,
+            PASSWORD: password,
+        })
+
+    def on_logout(self, username):
+        self.__send_alert(USER_LOGOUT_DESCRIPTION, {
+            USERNAME: username,
+        })
+
+    def on_file_sent(self, file):
+        self.__send_alert(USER_DOWNLOADED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(file),
+        })
+
+    def on_file_received(self, file):
+        self.__send_alert(USER_UPLOADED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(file),
+        })
+
+    # On the next section we override actual handlers instead of callbacks for more alerts
+    def ftp_LIST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_LIST(self, path)
+    
+    def ftp_NLST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_NLST(self, path)
+    
+    def ftp_MLST(self, path):
+        self.__send_alert(USER_LISTED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_MLST(self, path)
+    
+    def ftp_CWD(self, path):
+        self.__send_alert(USER_NAVIGATED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_CWD(self, path)
+    
+    def ftp_MKD(self, path):
+        self.__send_alert(USER_CREATED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_MKD(self, path)
+    
+    def ftp_RMD(self, path):
+        self.__send_alert(USER_DELETED_DIR_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_RMD(self, path)
+    
+    def ftp_DELE(self, path):
+        self.__send_alert(USER_DELETED_FILE_DESCRIPTION, {
+            ADDITIONAL_FIELDS: self.__format_file_path(path)
+        })
+        FTPHandler.ftp_DELE(self, path)
+
+
+class FTPAlertingServer(FTPServer):
+    def __init__(self, *args, **kwargs):
+        self.alerting_function = kwargs.pop("alerting_function")
+        self.base_dir = kwargs.pop("base_dir")
+        FTPServer.__init__(self, *args, **kwargs)
+
+
+class FTPService(ServerCustomService):
+    def __init__(self, *args, **kwargs):
+        super(FTPService, self).__init__(*args, **kwargs)
+        self.server = None
+        self.temp_dir = None
+
+    def prepare_temp_dir(self):
+        self.temp_dir = tempfile.mkdtemp()
```

Comment:
Consider using the dir param to have the temp dir created under your service's directory instead of somewhere that the OS is allowed to mess with and that someone could empty while the server is running.

Reply:
Not that important